bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt

```diff
@@ -698,22 +702,50 @@ private constructor(
         /**
          * Get the version string for this Mastodon instance.
          * @return a string corresponding to the version of this Mastodon instance
-         * @throws BigBoneRequestException if instance version can not be retrieved using any known method or API version
+         * @throws BigBoneClientInstantiationException if instance version cannot be retrieved using any known method or API version
          */
+        @Throws(BigBoneClientInstantiationException::class)
         private fun getInstanceVersion(): String {
-            try {
-                val serverInfoVersion = NodeInfoClient
-                    .retrieveServerInfo(instanceName)
-                    ?.software
-                    ?.takeIf { it.name == "mastodon" }
-                    ?.version
-                if (serverInfoVersion != null) return serverInfoVersion
-            } catch (_: BigBoneRequestException) {
+            return try {
+                getInstanceVersionViaServerInfo()
+            } catch (error: BigBoneClientInstantiationException) {
+                // fall back to retrieving from Mastodon API itself
+                try {
+                    getInstanceVersionViaApi()
+                } catch (instanceException: InstanceVersionRetrievalException) {
+                    throw BigBoneClientInstantiationException(
+                        message = "Failed to get instance version of $instanceName",
+                        cause = if (instanceException.cause == instanceException) {
+                            instanceException.initCause(error)
+                        } else {
+                            instanceException
+                        }
+                    )
+                }
             }
+        }
+
+        @Throws(ServerInfoRetrievalException::class)
+        private fun getInstanceVersionViaServerInfo(): String {
+            val serverSoftwareInfo: Server.Software? = NodeInfoClient
+                .retrieveServerInfo(instanceName)
+                ?.software
+                ?.takeIf { it.name == "mastodon" }
 
-            // fall back to retrieving from Mastodon API itself
-            val instanceVersion = getInstanceVersionFromApi(2) ?: getInstanceVersionFromApi(1)
-            return instanceVersion ?: throw BigBoneRequestException("Unable to fetch instance version")
+            if (serverSoftwareInfo != null) return serverSoftwareInfo.version
+
+            throw ServerInfoRetrievalException(
+                cause = IllegalArgumentException("Server $instanceName doesn't appear to run Mastodon")
+            )
+        }
+
+        @Throws(InstanceVersionRetrievalException::class)
+        private fun getInstanceVersionViaApi(): String {
+            return try {
+                getInstanceVersionFromApi(2)
+            } catch (e: InstanceVersionRetrievalException) {
+                getInstanceVersionFromApi(1)
+            }
```

Comment:
If we fail with the v2 version and then also fail with the v1 version, this would only throw the v1 exception to the caller. Adding the v2 exception to the v1 exception is not trivially possible as we likely already set a cause in the exception thrown in `getInstanceVersionFromApi`

Reply:
So do you suggest to remove v1 now here or to remove it later once you we officially no longer support v1?

Reply:
I would remove the v1 fallback separately, but consider it to be fine to have the drawback you describe right now.

Reply:
> Even if we keep the fallback of getting a version via API, we might want to remove the secondary "v1" fallback after all server versions still using it become unsupported at the end of the year. We don't need to make our code more complicated than it already is just some weeks before.

@bocops Where did you get this info from regarding end of year - just curious.

Reply:
https://github.com/mastodon/mastodon/releases/tag/v3.5.14

> ⚠️ The 3.5.x branch will not receive any update—including security fixes—after 2023-12-31.

Reply:
Thanks for the info @PattaFeuFeu . Well then, let's remove the v1 in a separate PR. Fine for me as well. :-)